aws_lambda_script/app/resources.py

```diff
@@ -0,0 +1,619 @@
+        ):
+            abort(406, description="Missing JSON data")
+
+        # Ensure the email is set to owner_email
+        updated_project["user"][0]["email"] = owner_email
+
+        data = read_data()
+        project = next(
+            (
+                proj
+                for proj in data["projects"]
+                if proj["details"]["name"] == project_name
+                and any(user["email"] == owner_email for user in proj["user"])
+            ),
+            None,
+        )
+
+        if not project:
+            abort(404, description="Project not found")
+
+        # Update the project details
+        project.update(updated_project)
+        write_data(data)
+
+        return project, 200
+
+
+# Read the client keys from S3 bucket for Cognito
+cognito_settings = read_client_keys()
+COGNITO_CLIENT_ID = cognito_settings["AWS_COGNITO_CLIENT_ID"]
+COGNITO_CLIENT_SECRET = cognito_settings["AWS_COGNITO_CLIENT_SECRET"]
+REDIRECT_URI = cognito_settings["REDIRECT_URI"]
+# REDIRECT_URI = 'http://localhost:8000/api/verify'
+
+verifyParser = reqparse.RequestParser()
+verifyParser.add_argument(
+    "code", location="args", required=True, help="Authorization code is required"
+)
+
+
+@ns.route("/verify")
+class VerifyToken(Resource):
+    # Route for handling the callback from Cognito
+    def get(self):
+        # Get the code from the query params
+        code = verifyParser.parse_args()["code"]
+        if not code:
+            logger.error("Authorization code not found")
+            return {"error": "Authorization code not found"}, 400
+
+        # Exchange the code for tokens
+        token_response = exchange_code_for_tokens(code)
+
+        if "id_token" not in token_response:
+            logger.error("Failed to retrieve ID Token")
+            return {"error": "Failed to retrieve ID Token"}, 400
+
+        return {
+            "id_token": token_response["id_token"],
+            "refresh_token": token_response["refresh_token"],
+        }, 200
+
+
+# First version of refreshing a token
+refreshParser = reqparse.RequestParser()
+refreshParser.add_argument(
+    "refresh_token", location="json", required=True, help="Refresh token is required"
+)
+
+
+@ns.route("/refresh")
+class RefreshToken(Resource):
+    def get(self):
+        refresh_token = refreshParser.parse_args()["refresh_token"]
+        if not refresh_token:
+            logger.error("Refresh token not found")
+            return {"error": "Refresh token not found"}, 400
+
+        token_response = exchange_refresh_token_for_id_token(refresh_token)
+
+        if "id_token" not in token_response:
+            logger.error("Failed to retrieve ID Token")
+            return {"error": "Failed to retrieve ID Token"}, 400
+
+        return {"id_token": token_response["id_token"]}, 200
+
+
+def exchange_refresh_token_for_id_token(refresh_token):
+    token_url = (
+        "https://keh-tech-audit-tool.auth.eu-west-2.amazoncognito.com/oauth2/token"
```

Comment:
These URLs may change, at a minimum we should make them a global definition.

Reply:
This won't change unless the name of the AWS cognito pool changes. I can add this into secret manager/S3/env if you want? I have made it global though to reduce duplicate code.